dbt-redshift/src/dbt/adapters/redshift/connections.py

```diff
@@ -517,37 +519,67 @@ def execute(
             table = agate_helper.empty_table()
         return response, table
 
-    def add_query(self, sql, auto_begin=True, bindings=None, abridge_sql_log=False):
-        connection = None
-        cursor = None
-
+    def add_query(
+        self, sql, auto_begin=True, bindings=None, abridge_sql_log=False
+    ) -> Tuple[Connection, Any]:
         self._initialize_sqlparse_lexer()
         queries = sqlparse.split(sql)
+        redshift_retryable_exceptions = (
+            redshift_connector.InterfaceError,
+            redshift_connector.InternalError,
+        )
 
-        for query in queries:
-            # Strip off comments from the current query
-            without_comments = re.sub(
-                re.compile(r"(\".*?\"|\'.*?\')|(/\*.*?\*/|--[^\r\n]*$)", re.MULTILINE),
-                "",
-                query,
-            ).strip()
-
-            if without_comments == "":
-                continue
-
-            retryable_exceptions = (
-                redshift_connector.InterfaceError,
-                redshift_connector.InternalError,
-            )
+        comment_regex = re.compile(r"(\".*?\"|\'.*?\')|(/\*.*?\*/|--[^\r\n]*$)", re.MULTILINE)
 
-            connection, cursor = super().add_query(
-                query,
-                auto_begin,
-                bindings=bindings,
-                abridge_sql_log=abridge_sql_log,
-                retryable_exceptions=retryable_exceptions,
-                retry_limit=self.profile.credentials.retries,
-            )
+        def _add_queries_with_retry(
+            queries_to_run: List[str],
+            retries: int,
+            backoff: int,
+            retry_exceptions: Tuple,
+            add_query_func: Callable[[Any, ...], Tuple[Connection, Any]],
+        ) -> Tuple[Connection, Any]:
+            try:
+                for query in queries_to_run:
+                    # Strip off comments from the current query
+                    without_comments = re.sub(
+                        comment_regex,
+                        "",
+                        query,
+                    ).strip()
+
+                    if without_comments == "":
+                        continue
+
+                    return add_query_func(
+                        query,
+                        auto_begin,
+                        bindings=bindings,
+                        abridge_sql_log=abridge_sql_log,
+                        retryable_exceptions=retry_exceptions,
+                        retry_limit=self.profile.credentials.retries,
+                    )
+
+            except redshift_connector.Error as e:
```

Comment:
Do we want to catch this exception here, or in `execute`? Doesn't this just queue the queries before (without) executing them?

Reply:
no that would be too obvious, instead `add_query` declares and calls `_execute_query_with_retry` within itself